src/main/controller/PromotionController.ts

```diff
@@ -183,12 +185,54 @@ export class PromotionController {
   ): Promise<any> => {
     try {
       await getManager().transaction(async (transactionalEntityManager) => {
-        const id = await IdValidation.schema.validateAsync(request.params.id, {
+        const pid = await IdValidation.schema.validateAsync(request.params.id, {
+          abortEarly: false,
+        });
+        const uid = await IdValidation.schema.validateAsync(request.body.uid, {
           abortEarly: false,
         });
+
+        await this.checkIfUserAndPromotionExist(
+          transactionalEntityManager,
+          pid,
+          uid
+        );
+
+        let voteRecord;
+        try {
+          voteRecord = await transactionalEntityManager
+            .getCustomRepository(VoteRecordRepository)
+            .findOneOrFail({ userId: uid, promotionId: pid });
+        } catch (e) {
+          // vote record does not exist
+          voteRecord = await transactionalEntityManager
+            .getCustomRepository(VoteRecordRepository)
+            .create({ userId: uid, promotionId: pid, voteState: VoteState.UP });
+          await transactionalEntityManager
+            .getCustomRepository(VoteRecordRepository)
+            .save(voteRecord);
+          await transactionalEntityManager
+            .getCustomRepository(PromotionRepository)
+            .increment({ id: pid }, 'votes', 1);
+          return response.status(204).send();
+        }
```

Comment:
How about this? I think it's a bit easier to read

```suggestion
        let voteRecord = await transactionalEntityManager
            .getCustomRepository(VoteRecordRepository)
            .findOne({ userId: uid, promotionId: pid });;
        if (!voteRecord) {
            await createVoteRecord(uid, pid, VoteState.UP, transactionalEntityManager)
            return response.status(204).send
        }
```

and actually, we can combine
```
          voteRecord = await transactionalEntityManager
            .getCustomRepository(VoteRecordRepository)
            .create({ userId: uid, promotionId: pid, voteState: VoteState.UP });
          await transactionalEntityManager
            .getCustomRepository(VoteRecordRepository)
            .save(voteRecord);

into this:

          await transactionalEntityManager
            .getCustomRepository(VoteRecordRepository)
            .save({ userId: uid, promotionId: pid, voteState: VoteState.UP }));

```

Reply:
Thanks Eric! What is `createVoteRecord` function ? Did you make?

Reply:
Oh no I didn't make it, but I meant like we can pull it out into its own method

Reply:
I see, thanks! I'll make clean the piece of code more!